### PR TITLE
Redact Entity

### PR DESCRIPTION
**Pull request by vivster7**

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/agora/server/business/AgoraBusiness.scala [hunk 1]
```diff
@@ -47,6 +47,23 @@ class AgoraBusiness {
...
+  def delete(agoraEntity: AgoraEntity, entityTypes: Seq[AgoraEntityType.EntityType], username: String): Int = {
+    if (!NamespacePermissionsClient.getNamespacePermission(agoraEntity, username).canRedact)
+      throw new NamespaceAuthorizationException(AgoraPermissions(Redact), agoraEntity, username)
+
+    // if the entity was a method, then redact all associated configurations
+    if (entityTypes equals AgoraEntityType.MethodTypes) {
+      val entityWithId = AgoraDao.createAgoraDao(entityTypes)
+        .findSingle(agoraEntity.namespace.get, agoraEntity.name.get, agoraEntity.snapshotId.get)
+
+      val configurations = AgoraDao.createAgoraDao(entityTypes).findConfigurations(entityWithId.id.get)
```
Comment (gbggrant) [inline src/main/scala/org/broadinstitute/dsde/agora/server/business/AgoraBusiness.scala]: Might be slightly more efficient to create the agora Dao once and re-use it later?
val agoraDao = AgoraDao.createAgoraDao(entityTypes)
agoraDao.find...

File: src/main/scala/org/broadinstitute/dsde/agora/server/business/AgoraBusiness.scala [hunk 2]
```diff
@@ -47,6 +47,23 @@ class AgoraBusiness {
...
+      configurations.foreach {config => AgoraEntityPermissionsClient.deleteAllPermissions(config)}
+    }
+
+    AgoraEntityPermissionsClient.deleteAllPermissions(agoraEntity)
+
```
Comment (gbggrant) [inline src/main/scala/org/broadinstitute/dsde/agora/server/business/AgoraBusiness.scala]: extra line.

File: src/main/scala/org/broadinstitute/dsde/agora/server/webservice/routes/RouteHelpers.scala [hunk 3]
```diff
@@ -113,11 +112,24 @@ trait QuerySingleHelper extends BaseRoute {
                               path: String,
                               queryHandler: Props): Unit = {
     addUserIfNotInDatabase(username)
-    val entityType = AgoraEntityType.byPath(path)
-    val message = QuerySingle(context, entity, entityType, username, onlyPayload)
+    val entityTypes = AgoraEntityType.byPath(path)
+    val message = QuerySingle(context, entity, entityTypes, username, onlyPayload)
 
-    perRequest(context, queryHandler, message, java.lang.Thread.currentThread.getStackTrace()(1).getMethodName)
+    perRequest(context, queryHandler, message)
```
Comment (gbggrant) [inline src/main/scala/org/broadinstitute/dsde/agora/server/webservice/routes/RouteHelpers.scala]: Why did this method argument go away?

**Comment by gbggrant** [comment]
:+1:

File: src/main/scala/org/broadinstitute/dsde/agora/server/webservice/routes/RouteHelpers.scala (hunk 3 above)
Comment (vivster7) [inline src/main/scala/org/broadinstitute/dsde/agora/server/webservice/routes/RouteHelpers.scala]: perRequest constructur defaults the method arg:

```scala
def perRequest(r: RequestContext, props: Props, message: AnyRef, name: String = java.lang.Thread.currentThread.getStackTrace()(1).getMethodName, timeout: Duration = 1.minutes)
```

**Pull request merged**